### PR TITLE
Add frequency and phase angle

### PR DESCRIPTION
**Pull request by SpecialMike**
# What does this implement/fix?

This adds frequency and phase angle reporting back to HA. Frequency is only reported for Phase A, while phase angle is reported for the other two phases. I'm not familiar with ESPHome enough to limit the configuration to make sure that `frequency` is only able to be defined for Phase A and `phase_angle` for the other two, so some pointers there would be appreciated. I found this through static analysis of a firmware dump.

As part of this, I also had to backport some changes from upstream. This repo doesn't seem to be a fork of the main esphome repo, so I wasn't sure if there was an easier way to get those changes to get this to work. These are the changes to:
- `esphome/components/sensor/__init__.py`
- `esphome/const.py`
- `.pre-commit-config.yaml`

I also had a lot of difficulty getting PlatformIO 5.x to work so I had to upgrade a few other files to get things to compile, but I did not include those changes with this PR.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
sensor:
  - platform: emporia_vue
    i2c_id: i2c_a
    phases:
      - id: phase_a  # Verify that this specific phase/leg is connected to correct input wire color on device listed below
        input: BLACK  # Vue device wire color
        calibration: 0.0230  # 0.022 is used as the default as starting point but may need adjusted to ensure accuracy
        # To calculate new calibration value use the formula <in-use calibration value> * <accurate voltage> / <reporting voltage>
        voltage:
          name: "Phase A Voltage"
          filters: [*moving_avg, *pos]
        frequency:
          name: "Phase A Frequency"
          filters: [*moving_avg, *pos]
      - id: phase_b  # Verify that this specific phase/leg is connected to correct input wire color on device listed below
        input: RED  # Vue device wire color
        calibration: 0.0215  # 0.022 is used as the default as starting point but may need adjusted to ensure accuracy
        # To calculate new calibration value use the formula <in-use calibration value> * <accurate voltage> / <reporting voltage>
        voltage:
          name: "Phase B Voltage"
          filters: [*moving_avg, *pos]
        phase_angle:
          name: "Phase B Phase Angle"
          filters: [*moving_avg, *pos]

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by clowrey** [comment]
Awesome! I really appreciate the work to get this integrated! will be useful to me thanks!

**Review by priitka — COMMENTED** [review]
Good stuff, only
` 
uint16_t raw_phase_angle = sensor_reading.degrees[((uint8_t)this->input_wire_)-1];
`
this->input_wire_ - 1 is -1 for BLACK.  I suggest to add check and set degrees to 0 for BLACK wire

**Comment by SpecialMike** [comment]
I also found the exact values for the current value correction, so I added that to this PR.

**Comment by SpecialMike** [comment]
Good catch, @priitka. That's fixed now with a simple check for both frequency and phase angle. Now if it is added to the configuration for an unsupported input phase, nothing will be reported.

I also should have mentioned in the PR that this calculation shows exactly what was being sent over MQTT, but it doesn't seem to match my real frequency here. I tested with a multimeter and found that my AC frequency is (correctly) 60Hz, whereas the emporia is reporting 61.1 Hz. I don't have any real way of testing this reading, so an additional correction may be needed. Any additional testing on this would be helpful to figure out the real relationship between what the sensors report and the real frequency.

The Phase angle is correctly reported as 180 degrees though.

**Comment by SpecialMike** [comment]
I just noticed that the existing power `correction_factor`s are close to the values I found in the firmware used for power and current calculations (`1/5.5 = 0.1818..`, `1/22 = 0.04545...`). I found that `775.0d/42624.0d` (0.018182244...) and `775.0d/170496.0d` (0.45455...) produce the same exact double as found in the firmware (they multiply this by 10 for power). Do these constants mean anything? Having the same numerator makes me think that they are correct.

**Comment by priitka** [comment]
I have here 3 phase, 230V 50Hz system.
Raw values reported by vue2 are frequency: 507 degrees: 338 and 173
I've measured with oscillograph that real frequency is indeed 50Hz, drifting slowly between 49.976Hz and 50.031Hz. So for me 25350.0 / sensor_reading.frequency would produce correct reading.
Degrees swapped though, 2nd phase should be 120 degree and 3rd 240.  338 / 507 * 360 = 240 and 173 / 507 * 360 = 122 -- close enough, but swapped

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,16 +1,17 @@
+---
```
Comment (flaviut) [inline .pre-commit-config.yaml]: can you remove this file from the changeset? I've pulled the latest esphome into dev and replaced this MR on top of that.

File: esphome/components/emporia_vue/sensor.py [hunk 2]
```diff
@@ -96,6 +101,18 @@
                         state_class=STATE_CLASS_MEASUREMENT,
                         accuracy_decimals=1,
                     ),
+                    cv.Optional(CONF_FREQUENCY): sensor.sensor_schema(
```
Comment (flaviut) [inline esphome/components/emporia_vue/sensor.py]: I think this validation could be made more precise by doing `cv.Any(...)`, with the nested config being either an phase angle-configurable phase, or a frequency-configurable phase.

**Review by flaviut — COMMENTED** [review]
Nice work here, thank you for taking the time to make and submit this contribution!

I'd be comfortable merging the current part of this right now with the small tweaks outlined below, but I'd like to get a definitive constant for the frequency. Since the power company very precisely sets the frequency, can you capture raw data for 24 hours, average it, and use that to come up with the calibration constant?

**Comment by SpecialMike** [comment]
Yeah, I didn't want to put the current change in this PR, but the pre-commit hook was giving me trouble. I'll submit a different PR in a bit for that and keep the frequency change in here.

**Comment by priitka** [comment]
Unrelated question to this pull request: what do you use for reversing esp32 binaries? 
I've tried Ghidra with ghidra-xtensa extension, but I get lots of /* WARNING: Control flow encountered unimplemented instructions */. Is there better disassembler for esp32?

**Comment by SpecialMike** [comment]
I'm using that extension as well, but I used the flash loader as well in an older version (9.1?) and then upgraded to the latest version. I fixed a few things in ghidra-xtensa to get argument passing for arguments that are larger than one word, which let me see where the doubles they are using were going. Not sure why they would use a double when a float would (I think) have enough precision for this. 

The fix I did is a bit of a hack, and you end up with a lot of `extra_out` variables in the decompilation. I may submit a PR over there later when I do a little more work on it. Ghidra really doesn't seem compatible with the Xtensa ISA, so I'm not sure support will ever get a lot better.

**Comment by SpecialMike** [comment]
I just remembered that I had to modify the flash loader extension for ghidra as well in 9.1 to get it to work, but I think I lost that change. :(

**Comment by priitka** [comment]
For me raw frequency has been constantly 507 for several days.
I'd be useful to get raw values from other grid frequencies to work out constant

Degrees drift a bit degrees[0] 332 - 339,  degrees[1] 169 - 172
Values are within acceptable range, only question is order of them. 
It makes sense if internally:
degrees[0] is timeunits from beginning of phase2 sine wave start till beginning of next phase 1 sine wave start 
degrees[1] is timeunits from beginning of phase3 sine wave start till beginning of next phase 1 sine wave start 
frequency is timeunits from beginning of phase1 sine wave start till beginning of next phase 1 sine wave start 

so 
phase2_degrees = 360 * (frequency-degrees[0]) / frequency
phase3_degrees = 360 * (frequency-degrees[1]) / frequency

**Comment by flaviut** [comment]
Wow, that's some great work on the ghirda plugin. It would be awesome to see improvements there made public.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (SpecialMike) [reply .pre-commit-config.yaml]: Resolved with 2d665e8e57f9c39d43d03a8385f3943d80de7e7b

**Comment by dschall** [comment]
What's the status on this?
I just installed my Vue, and I'd love to get frequency and phase angle to show on my smart home.

File: esphome/components/emporia_vue/sensor.py (hunk 2 above)
Comment (SpecialMike) [reply esphome/components/emporia_vue/sensor.py]: I read through the docs for `voluptuous` and think that `cv.Any` is only for value validation. I can't find any way of saying that a configuration can have only one or the other key. I'm not very familiar with the library, so I could be wrong on that.

**Comment by SpecialMike** [comment]
Sorry for the delay.

The frequency reported by my device has been 61.1Hz the entire time, though the phase angle has varied a little over time (is that normal?). For @priitka, it sounds like the reading was exactly correct. I'm not sure if this is a hardware issue with my unit, or if I'm missing something else in the firmware, but I don't really see how it would be different for 50Hz vs 60Hz. I'll probably add a `muliply` filter to my configuration, and we should recommend others do the same to calibrate it, unless it's worth making it a configuration setting like the voltage for the phases.

**Comment by dschall** [comment]
Thanks, I'd be happy to test this on my platform (USA, 2-phase 60 Hz) and report the results.
I just don't know how to build with these changes, any pointers?

**Comment by flaviut** [comment]
See https://github.com/emporia-vue-local/esphome/discussions/109#discussioncomment-4692785 for instructions.

I'll follow this up later this week/weekend.

**Comment by dschall** [comment]
Thanks for the pointers.
I rebased your PR onto the most recent commit in dev and tried it out.
Looks like both frequency and phase are reporting for me, have plausible values and show little variance (which I guess means, it's working):
image


image

**Comment by dschall** [comment]
One issue to point out:
The phase angle is reported as "frequency" in the config:
image

**Comment by dschall** [comment]
PS: My frequency measurement is not accurate. I've double-checked with a precision multimeter, and it's showing as 60Hz precisely vs. the Vue showing 61.7Hz.
So some calibration factor may be necessary on that measurement.

PS: I'm using various websites to double-check my frequency measurements [1] [2] and since the frequency for the entire US is synchronized, I think we can rely on these values for individual homes, too.
I'm currently correcting the value by adding a filter on the frequency:
```
  - &adjust_freq
    # Adjust the value by a correction factor
    lambda: 'return x * 0.97235008103727714748784440842786f;'
```

[1] https://fnetpublic.utk.edu/frequencymap.html
[2] https://www.ercot.com/content/cdr/html/real_time_system_conditions.html

**Comment by flaviut** [comment]
Testing on my 60 Hz system for a few hours shows a average of 421.5008 raw frequency counts. I've updated the code to use this new constant, which becomes a very clear `raw_frequency / 7.025f`.

I've also added validation to ensure that the config is valid at the build stage rather than at runtime.

**Comment by flaviut** [comment]
Merged via https://github.com/emporia-vue-local/esphome/commit/47c623d7040bcd4da0cc3ae98c71499cd3e2b6e0 :)

Thank you for implementing this, it'll be very useful!

**Comment by dschall** [comment]
Thanks for merging this.
Just one question, out of curiosity:
How do you know it's `raw_frequency / const` and not `some other const / raw_frequency`?

**Comment by flaviut** [comment]
That was a very polite way to point out to me that I'd made a mistake :)

I assumed that increasing values means increasing frequency, but it is actually the reverse is true. Seems like `25310` is a more correct constant--calculated averaging the 50Hz raw value from above with my own raw value, and then rounding:

|raw|calculated|actual|
|---|---|---|
|421.5|60.04744958|60|
|507|49.92110454|50|

Corrected in d7cc89375579c014a3e82b94f9c1cb1ad9ac3542.

**Comment by dschall** [comment]
Haha, I wish I was that polite ;)
I was really just wondering if you had found the calculation in the original binary or how you came to this conclusion.

To me, it also looks strange that the inverse is correct here: `const / raw_freq`.
Why would the raw value go _up_ when the frequency goes _down_ (and vice versa)? I'm not into electronics, so to my layman's eye, it looks confusing..

And finally, this may need a calibration factor, because the frequency still doesn't quite add up for me.
I'll double check with your changes.

**Comment by dschall** [comment]
Well, guess it's as good as it gets:
showing 60.0 Hz, when the actual frequency is about 59.975.

**Comment by flaviut** [comment]
> Why would the raw value go up when the frequency goes down (and vice versa)? I'm not into electronics, so to my layman's eye, it looks confusing.

Perhaps there's a timer that counts up every so many milliseconds and then gets captured and reset whenever the wave crosses 0?

>  showing 60.0 Hz, when the actual frequency is about 59.975.

Yup. The raw data only has about a quarter digit of precision. Even one digit after the decimal is sort of implying that there's more precision here than there actually is.

**Comment by dschall** [comment]
>Perhaps there's a timer that counts up every so many milliseconds and then gets captured and reset whenever the wave crosses 0?

Makes sense.

Btw, I got two devices, because I have another subpanel, and the frequency measurement does not appear to be very reliable.
I hooked them both into the same panel, same phases for testing, and one is showing 60 Hz, while the second one is reading 59.7 Hz.

**Comment by flaviut** [comment]
> the frequency measurement does not appear to be very reliable. I hooked them both into the same panel, same phases for testing, and one is showing 60 Hz, while the second one is reading 59.7 Hz.

This is what I tried to get across with my previous comment. These results are reliable.

The problem is the resolution is low. The system can never read 59.8 or 59.9. It's either 60.0 or 59.7. The fact that it's only off by step from the true value is good!

**Comment by dschall** [comment]
Is that a fundamental limitation of the Vue?
Since the frequency in both Europe and the US is pretty stable and usually fluctuates by less than 0.1 Hz, having a 0.25 Hz resolution at best is absolutely worthless..

**Comment by flaviut** [comment]
> Is that a fundamental limitation of the Vue?

Yes: the hardware could support more precise results, but the firmware on the SAM microcontroller doesn't.

> Since the frequency in both Europe and the US is pretty stable and usually fluctuates by less than 0.1 Hz, having a 0.25 Hz resolution at best is absolutely worthless..

Yup, for those of us with good power.

Some folks in places with unstable power grids mentioned they could use it to file claims with their electric providers if their equipment was damaged. 

Perhaps it could also be used to warm up generators if it indicates an imminent power outage.

**Pull request closed without merging**